### PR TITLE
Add helm chart for the operator

### PR DESCRIPTION
**Pull request by zeenix**
This is the source for the Helm chart. You can use:

helm package helm/habitat-operator

command to create a chart package for the operator for distribution.

This chart uses the docker image of the latest release: 0.4.0.

### CODE REVIEW

File: helm/habitat-operator/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+$ helm install habitat/habitat-operator
+```
+
+## Introduction
+
+This chart bootstraps a [habitat-operator](https://github.com/kinvolk/habitat-operator) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
+
+## Prerequisites
+  - Habitat `>= 0.36.0`
+  - Kubernetes cluster with version `1.7.x`, `1.8.x` or `1.9.x`.
```
Comment (asymmetric) [inline helm/habitat-operator/README.md]: Can't we just say ">- 1.7.x". The above approach doesn't scale ;)

File: helm/habitat-operator/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+If role-based access control (RBAC) is enabled in your cluster, you may need to give Tiller (the server-side component of Helm) additional permissions. **If RBAC is not enabled, be sure to set `rbacEnable` to `false` when installing the chart.**
+
+1. Create a ServiceAccount for Tiller in the `kube-system` namespace
+```console
+$ kubectl -n kube-system create sa tiller
+```
+
+2. Create a ClusterRoleBinding for Tiller
+```console
+# kubectl v1.5.x
```
Comment (asymmetric) [inline helm/habitat-operator/README.md]: I think you should turn this into a section, e.g. `###`, and have two separate code blocks for each version of k8s.

File: helm/habitat-operator/README.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+```console
+$ helm delete my-release
+```
+
+The command removes all the Kubernetes components associated with the chart and deletes the release.
+
+## Configuration
+
+The following tables lists the configurable parameters of the habitat-operator chart and their default values.
```
Comment (asymmetric) [inline helm/habitat-operator/README.md]: *table

File: helm/habitat-operator/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+description: A Kubernetes controller designed to solve running and auto-managing Habitat services on Kubernetes.
+engine: gotpl
+maintainers:
+  - name: Zeeshan Ali
+    email: zeeshan@kinvolk.io
+name: habitat-operator
+sources:
+  - https://github.com/kinvolk/habitat-operator
+version: 0.4.0
```
Comment (LiliC) [inline helm/habitat-operator/Chart.yaml]: Where did this version come from? Considering this is the first release of this chart this version number should be `0.0.1`?

File: helm/habitat-operator/Chart.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+description: A Kubernetes controller designed to solve running and auto-managing Habitat services on Kubernetes.
```
Comment (LiliC) [inline helm/habitat-operator/Chart.yaml]: Nit: Would get rid of the "auto" part here. Would add creating as well.

File: helm/habitat-operator/README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
+# habitat-operator
+
+Installs [habitat-operator](https://github.com/kinvolk/habitat-operator) to create/configure/manage Habitat services atop Kubernetes.
+
+## TL;DR;
+
+```console
+$ helm repo add habitat https://urltochart.org/
```
Comment (LiliC) [inline helm/habitat-operator/README.md]: urltochart.org? :)

File: helm/habitat-operator/README.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
+# habitat-operator
+
+Installs [habitat-operator](https://github.com/kinvolk/habitat-operator) to create/configure/manage Habitat services atop Kubernetes.
```
Comment (LiliC) [inline helm/habitat-operator/README.md]: s/atop/inside/g

File: helm/habitat-operator/README.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+## TL;DR;
+
+```console
+$ helm repo add habitat https://urltochart.org/
+$ helm install habitat/habitat-operator
+```
+
+## Introduction
+
+This chart bootstraps a [habitat-operator](https://github.com/kinvolk/habitat-operator) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
```
Comment (LiliC) [inline helm/habitat-operator/README.md]: s/on a/inside of/g

File: helm/habitat-operator/README.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+## Prerequisites
```
Comment (LiliC) [inline helm/habitat-operator/README.md]: I wouldn't duplicate these, maybe just link in the operator README section?

File: helm/habitat-operator/README.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+The command removes all the Kubernetes components associated with the chart and deletes the release.
+
+## Configuration
+
+The following tables lists the configurable parameters of the habitat-operator chart and their default values.
+
+Parameter | Description | Default
+--- | --- | ---
+`image.repository` | Image | `docker.io/kinvolk/habitat-operator`
+`image.tag` | Image tag | `v0.4.0`
```
Comment (LiliC) [inline helm/habitat-operator/README.md]: Can you please add this to the https://github.com/kinvolk/habitat-operator/blob/master/Makefile#L30 So that it is auto updated when doing the release.

File: helm/habitat-operator/templates/NOTES.txt [hunk 11]
```diff
@@ -0,0 +1,4 @@
+The Habitat Operator has been installed. Check its status by running:
+  kubectl --namespace {{ .Release.Namespace }} get pods -l "app={{ template "habitat-operator.name" . }},release={{ .Release.Name }}"
+
+You can now deploy Habitat services through Helm.
```
Comment (LiliC) [inline helm/habitat-operator/templates/NOTES.txt]: Think you do not need Helm to deploy Habitat services, maybe rephrase this?

File: helm/habitat-operator/templates/clusterrole.yaml [hunk 12]
```diff
@@ -0,0 +1,47 @@
+{{- if .Values.rbacEnable }}
+{{- if .Capabilities.APIVersions.Has "rbac.authorization.k8s.io/v1beta1" }}
```
Comment (LiliC) [inline helm/habitat-operator/templates/clusterrole.yaml]: Hmm but now RBAC is not in beta or alpha but it is v1. Maybe add that condition as well?

File: helm/habitat-operator/templates/clusterrolebinding.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
+{{- if .Values.rbacEnable }}
+{{- if .Capabilities.APIVersions.Has "rbac.authorization.k8s.io/v1beta1" }}
```
Comment (LiliC) [inline helm/habitat-operator/templates/clusterrolebinding.yaml]: Same as above for RBAC.

File: helm/habitat-operator/templates/deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        app: {{ template "habitat-operator.name" . }}
+        operator: habitat
+        release: {{ .Release.Name }}
+    spec:
+      containers:
+        - name: {{ template "habitat-operator.name" . }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: "{{ .Values.image.pullPolicy }}"
+          ports:
+            - containerPort: 8080
```
Comment (LiliC) [inline helm/habitat-operator/templates/deployment.yaml]: A bit confused by this? Why is this needed?

File: helm/habitat-operator/values.yaml [hunk 15]
```diff
@@ -0,0 +1,25 @@
+## Habitat-operator image
+##
+image:
+  repository: docker.io/kinvolk/habitat-operator
+  tag: v0.4.0
```
Comment (LiliC) [inline helm/habitat-operator/values.yaml]: Same as above, can you please add this to the https://github.com/kinvolk/habitat-operator/blob/master/Makefile#L30 So that it is auto updated when doing the release.

File: helm/habitat-operator/README.md (hunk 1 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: sure

File: helm/habitat-operator/README.md (hunk 2 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: Good idea. Actually this confused me in the prometheus chart and I ended up using the incorrect command.

File: helm/habitat-operator/Chart.yaml (hunk 5 above)
Comment (zeenix) [reply helm/habitat-operator/Chart.yaml]: hmm..creating? The operator doesn't create them, does it? To me the "designed to solve" sounds more redundant. How about:

"A Kubernetes controller that manages Habitat services on Kubernetes."

File: helm/habitat-operator/Chart.yaml (hunk 4 above)
Comment (zeenix) [reply helm/habitat-operator/Chart.yaml]: Like many other things pointed out in reviews here, a leftover from copy&paste of prometheus operator chart. :)

File: helm/habitat-operator/README.md (hunk 7 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: since it doesn't say "Kubernetes cluster", "atop" is correct and sounds better to me so I kept that  from the prometheus operator chart.

File: helm/habitat-operator/README.md (hunk 6 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: exactly! we don't have a URL yet. :) Once we know where to host it, we can set this.

File: helm/habitat-operator/README.md (hunk 8 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: hmm.. in here it does say "cluster" so that would be more accurate but I don't think "on" is incorrect either and I'd prefer it unless you feel strongly about this.

File: helm/habitat-operator/README.md (hunk 9 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: you mean link *to* the operator README? sure.

File: helm/habitat-operator/README.md (hunk 10 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: Sure thing but we really shouldn't be modifying source files. The convention is to have put the source content into .in files and then build can generate the final version.

File: helm/habitat-operator/templates/NOTES.txt (hunk 11 above)
Comment (zeenix) [reply helm/habitat-operator/templates/NOTES.txt]: ok

File: helm/habitat-operator/templates/clusterrole.yaml (hunk 12 above)
Comment (zeenix) [reply helm/habitat-operator/templates/clusterrole.yaml]: c&p mistake, I'll just modify it to v1 everywhere.

File: helm/habitat-operator/templates/deployment.yaml (hunk 14 above)
Comment (zeenix) [reply helm/habitat-operator/templates/deployment.yaml]: another c&p mistake probably. Will remove or come back with reasons (if any).

File: helm/habitat-operator/README.md (hunk 9 above)
Comment (LiliC) [reply helm/habitat-operator/README.md]: Yes, link to to the operator README.

File: helm/habitat-operator/README.md (hunk 10 above)
Comment (LiliC) [reply helm/habitat-operator/README.md]: Sure, feel free to do that.

File: helm/habitat-operator/templates/clusterrole.yaml (hunk 12 above)
Comment (LiliC) [reply helm/habitat-operator/templates/clusterrole.yaml]: No, I meant this is good, as we are supporting k8s versions of 1.7 and 1.8 as well. But we need to add v1 as well for 1.9+.

File: helm/habitat-operator/README.md (hunk 10 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: hmm.. I already modified the Makefile rule for existing yaml files but the problem with generating README.md is that then you don't get the nice integration with github. :( Probably best not the mention the actual version here.

**Comment by zeenix** [comment]
PR updated. For the installation, let's for now go with the simplest approach of providing chart tarball as one of the release artifacts. We can look into moving the chart source and tarball to another repo (possibly the main charts repo) later.

**Comment by zeenix** [comment]
Ok, realized that we have to setup a chart repo for being able to add operator as a dep to other charts. I update the PR again and chose a canonical location on kinvok.io where we'll put helm charts for now and will be our repo for now.

I also removed my patch about version update and now version is updated the same as before.

**Comment by zeenix** [comment]
Alright so now even the helm chart repo is setup and I tested the instruction to work. Now would be the best time to PATALA and merge.

**Comment by asymmetric** [comment]
PATALA? 😂

File: Makefile [hunk 16]
```diff
@@ -28,7 +28,13 @@ clean-test:
 	kubectl delete clusterrole habitat-operator
 
 update-version:
-	find examples -name "*.yml" -type f -exec sed -i.bak "s/habitat-operator:.*/habitat-operator:v$$(cat VERSION)/g" '{}' \;
-	find examples -name "*.yml.bak" -type f -exec rm '{}' \;
+	find examples helm -name "*.yml" -o -name "*.yaml" -type f \
```
Comment (asymmetric) [inline Makefile]: Can't we do `find . -regex '.*\.ya?ml$'`?

Also, the sed invocation was changed quite a bit. Why? I think this could've been its own commit.

File: helm/habitat-operator/Chart.yaml (hunk 4 above)
Comment (asymmetric) [reply helm/habitat-operator/Chart.yaml]: So will this be fixed or should it stay like this?

**Comment by blixtra** [comment]
Perhaps it's good to point out how the charts are hosted. Right now the instructions for hosting are only on the gh-pages branch. Wouldn't hurt to at least point to that from the main documentation.
wdyt, @LiliC @asymmetric?

File: Makefile (hunk 16 above)
Comment (zeenix) [reply Makefile]: > Can't we do find . -regex '.*\.ya?ml$'?

That's hard to read so I'll be against it.

> Also, the sed invocation was changed quite a bit. Why?

too vague a question. :) The files in the chard are "yaml" and in examples they're "yml", we need to tell it to look into another directory as well. We could add a separate sed line for helm part.

> I think this could've been its own commit.

I don't think so but I don't have strong feelings about it so if you have, I can split it.

File: helm/habitat-operator/Chart.yaml (hunk 4 above)
Comment (zeenix) [reply helm/habitat-operator/Chart.yaml]: after a bit of thinking and discussion on slack, I decided to keep the version in sync. It's easier to manage and the distributed charts won't cause any confusions.

Comment (zeenix) [reply helm/habitat-operator/Chart.yaml]: I mean the chart and the operator versions.

**Comment by asymmetric** [comment]
@blixtra Since this PR introduces a Helm specific README, I think that info could go there.

**Comment by zeenix** [comment]
@asymmetric way ahead of you. :) Info already added to that README. :)

File: Makefile [hunk 17]
```diff
@@ -28,7 +28,13 @@ clean-test:
...
+		-exec sed -i.bak \
+		-e "s/habitat-operator:.*/habitat-operator:v$$(cat VERSION)/g" \
```
Comment (asymmetric) [inline Makefile]: What are the `-e` flags doing? Is it for executing multiple substitutions in one sed run?

File: helm/habitat-operator/templates/clusterrole.yaml (hunk 12 above)
Comment (asymmetric) [reply helm/habitat-operator/templates/clusterrole.yaml]: What should we do here? Or is this good?

File: helm/habitat-operator/templates/clusterrolebinding.yaml (hunk 13 above)
Comment (asymmetric) [reply helm/habitat-operator/templates/clusterrolebinding.yaml]: Same, needs to be fixed?

File: Makefile (hunk 17 above)
Comment (zeenix) [reply Makefile]: -e is for specifying expressions but it can be (and is usually) omitted if you only have one expression.

File: helm/habitat-operator/templates/clusterrole.yaml (hunk 12 above)
Comment (zeenix) [reply helm/habitat-operator/templates/clusterrole.yaml]: I think so. I added the v1 case below.

File: helm/habitat-operator/templates/clusterrolebinding.yaml (hunk 13 above)
Comment (zeenix) [reply helm/habitat-operator/templates/clusterrolebinding.yaml]: same answer here too. :)

File: Makefile (hunk 16 above)
Comment (asymmetric) [reply Makefile]: But `'.*.ya?ml$'?` is a regexp and regexps are 1337.

Comment (zeenix) [reply Makefile]: No :)

File: helm/habitat-operator/README.md [hunk 18]
```diff
@@ -0,0 +1,109 @@
+# habitat-operator
+
+Installs [habitat-operator](https://github.com/kinvolk/habitat-operator) to manage Habitat services in a Kubernetes cluster.
+
+## TL;DR;
+
+```console
+$ helm repo add habitat https://kinvolk.github.io/habitat-operator/helm/charts/stable/
+$ helm install habitat/habitat-operator
```
Comment (LiliC) [inline helm/habitat-operator/README.md]: Is this supposed to work already?

File: helm/habitat-operator/templates/clusterrole.yaml (hunk 12 above)
Comment (LiliC) [reply helm/habitat-operator/templates/clusterrole.yaml]: Yes, this is fine now!

File: helm/habitat-operator/README.md (hunk 18 above)
Comment (zeenix) [reply helm/habitat-operator/README.md]: Yeah, I tested it here. If you have helm setup, please do try it out too.

File: helm/habitat-operator/README.md [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+```console
+$ helm install --name my-release habitat/habitat-operator -f values.yaml
+```
+
+> **Tip**: You can use the default [values.yaml](values.yaml)
+
+## Hosting
+
+The helm chart repository is managed on the [`gh-pages` branch](https://github.com/kinvolk/habitat-operator/tree/gh-pages) and instructions for repository management can be found [here](https://github.com/kinvolk/habitat-operator/tree/gh-pages/helm/charts).
```
Comment (LiliC) [inline helm/habitat-operator/README.md]: Nitpick warning: Capitalize Helm ;)

**Comment by zeenix** [comment]
nitpick addressed. :)
